src/app/hooks/useTasks.ts

```diff
@@ -1,57 +1,84 @@
-import { useState, useEffect } from "react";
 import { Task } from "../types/task";
+import useSWR, { mutate } from "swr";
+import { useState } from "react";
 
-export function useTasks(userId: string | undefined) {
-  const [tasks, setTasks] = useState<Task[]>([]);
-
-  useEffect(() => {
-    if (userId) {
-      fetchTasks();
-    }
-  }, [userId]);
+const fetcher = async (url: string) => {
+  const response = await fetch(url);
+  if (!response.ok) {
+    throw new Error("Failed to fetch tasks");
+  }
+  return response.json();
+};
 
-  const fetchTasks = async () => {
-    try {
-      const response = await fetch("/api/listTasks");
-      if (!response.ok) {
-        throw new Error("Failed to fetch tasks");
-      }
-      const data = await response.json();
-      setTasks(data);
-    } catch (error) {
-      console.error("Failed to fetch tasks:", error);
-    }
-  };
+export function useTasks(userId: string | undefined) {
+  const { data: tasks = [], error } = useSWR<Task[]>(
+    userId ? "/api/list-tasks" : null,
+    fetcher
+  );
+  const [isAdding, setIsAdding] = useState(false);
 
   const addTask = async (formData: FormData) => {
+    setIsAdding(true);
     try {
-      const response = await fetch("/api/createTask", {
+      const formDataTask = {
+        title: formData.get("title") as string,
+        description: formData.get("description") as string,
+        category: formData.get("category") as string,
+        date: formData.get("date") as string,
+        userId: formData.get("userId") as string,
+      };
+
+      mutate("/api/list-tasks", [...(tasks || []), formDataTask], false);
+
+      const response = await fetch("/api/create-task", {
         method: "POST",
         body: formData,
       });
+
       if (!response.ok) {
         throw new Error("Failed to add task");
       }
+
       const newTask = await response.json();
-      setTasks((prevTasks) => [...prevTasks, newTask]);
+
+      mutate("/api/list-tasks");
     } catch (error) {
       console.error("Failed to add task:", error);
+      mutate("/api/list-tasks");
+    } finally {
+      setIsAdding(false);
     }
   };
 
   const deleteTask = async (id: string) => {
     try {
-      const response = await fetch(`/api/deleteTask/${id}`, {
+      mutate(
+        "/api/list-tasks",
+        tasks.filter((task) => task.id !== id),
```

Comment:
**Thought: when you mutate data you can call useSWR#mutate that will blow the cache for other requests that depend on it**